snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/event/Event.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2015-2020 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.event;
+
+// Java
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import java.util.List;
+import java.util.Map;
+
+// This library
+import com.snowplowanalytics.snowplow.internal.tracker.Tracker;
+import com.snowplowanalytics.snowplow.payload.Payload;
+import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;
+
+/**
+ * The event interface
+ */
+public interface Event {
+
+    /**
+     * @return the event custom contexts
+     */
+    @NonNull List<SelfDescribingJson> getContexts();
+
+    /**
+     * @return the optional true events timestamp
+     */
+    @Nullable
+    Long getTrueTimestamp();
+
+    /**
+     * @return the event data payload
+     */
+    @NonNull Map<String, Object> getDataPayload();
+
+    /**
+     * Hook method called just before the event processing in order to execute special operations.
+     *
+     * @apiNote Internal use only - Don't use in production, it can change without notice.
+     */
+    void beginProcessing(@NonNull Tracker tracker);
```

Comment:
The main issue is that those methods expose the Tracker class which is private (at least, that is the intention in the future). That wouldn't be a blocker as we could expose the tracker with a simple interface. They exist because I needed to fix a problem with some legacy events but probably they wouldn't be needed with a different event design.
ECommerce transaction event uses endProcessing to post an event for each item in the basket. I think they should be just contexts/entities attached to the transaction event rather than a stream of events.
Instead, for the GDPR events (ConsentGranted) we could get rid of beginProcessing as the documents could be added directly in the customContexts.
For this reason I wonder if we should keep them alive in the future.
My doubt is: do they solve a problem that is not solvable in a more elegant way?

Reply:
The reason I suggested this as it is quite similar to what I've done with the plugins in the JS Tracker. I've got a Tracker interface that I've commited to even though the tracker is an internal class (it's down in the browser-tracker-core package rather than in browser-track). However, you can get to this tracker interface via the plugins - this allows me to extend / make things optional, not something you're aiming for here but offering points of extension is an interesting idea for our users.
I'm not too bothered if we leave this marked as internal for v2 but it would be an interesting idea to open up opportunities to get at some limited internals for "advanced use cases".